crypto/threads_pthread.c

```diff
@@ -655,6 +835,28 @@ int CRYPTO_THREAD_unlock(CRYPTO_RWLOCK *lock)
 # ifdef USE_RWLOCK
     if (pthread_rwlock_unlock(lock) != 0)
         return 0;
+#  ifdef REPORT_RWLOCK_CONTENTION
+    {
+        struct stack_traces *traces = CRYPTO_THREAD_get_local(&thread_contention_data);
+
+        if (contention_fp == NULL)
+            return 1;
+        if (traces != NULL) {
+            /*
+             * We often in openssl have code points where we unlock a lock
+             * that wasn't locked in the first place.  pthread_rwlock_unlock
```

Comment:
What!!!??? I find this hard to believe. If so - then that's a bug that needs fixing.

Reply:
I'm referring here to https://github.com/openssl/openssl/pull/24779, in which two such call sites were fixed.  It appears there are more of them.

Reply:
sorry, I just realized this was being caused by the fact that I'm incrementing the lock_depth conditionally on blocking in read and write lock, but decrementing it unilaterally in unlock, I'll fix that and the added assert will catch any additional #24779 style bugs we may have